File: C++/utils/Serializer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <memory>
#include <cstring>
#include <stdexcept>
#include <functional>

#include "ByteBuffer.cpp"
#include "ByteReader.cpp"

// Base class for serializable objects
// Abstract base class for objects that can be serialized/deserialized
class JavaSerializable
{
public:
    virtual ~JavaSerializable() = default;
    virtual std::string getJavaClassName() const = 0;
    virtual void deserializeFields(ByteReader &reader) = 0;
    virtual std::vector<std::pair<std::string, std::string>> getFieldMetadata() const = 0;
    virtual void serializeFields(ByteBuffer &buffer) const = 0;
};

class JavaSerializer
{
private:
    static std::map<const void *, int> serializedObjects;
    static int objectCounter;

public:
    static std::vector<uint8_t> serialize(const JavaSerializable *obj)
    {
        serializedObjects.clear();
        objectCounter = 0;
        ByteBuffer buffer;
        serializeObject(obj, buffer);
        return buffer.getBuffer();
    }

private:
    static void serializeObject(const JavaSerializable *obj, ByteBuffer &buffer)
    {
        if (obj == nullptr)
        {
            buffer.writeByte(0); // null marker
            return;
        }

        buffer.writeByte(1); // non-null marker

        // Check for circular reference
        auto it = serializedObjects.find(obj);
        if (it != serializedObjects.end())
        {
            buffer.writeByte(1); // reference marker
            buffer.writeInt(it->second);
            return;
        }

        buffer.writeByte(0); // new object marker
        serializedObjects[obj] = objectCounter++;

        // Write class metadata
        std::string className = obj->getJavaClassName();
        buffer.writeString(className);

        // Write field count and metadata
        auto fields = obj->getFieldMetadata();
        buffer.writeInt(static_cast<int32_t>(fields.size()));

        // Write field metadata
        for (const auto &field : fields)
        {
            buffer.writeString(field.first);  // field name
            buffer.writeString(field.second); // field type
        }

        // Write field values
        obj->serializeFields(buffer);
    }
};

class ObjectFactory
{
private:
    static std::map<std::string, std::function<std::shared_ptr<JavaSerializable>()>> creators;

public:
    template <typename T>
    static void registerClass(const std::string &className)
    {
        creators[className] = []()
        { return std::make_shared<T>(); };
    }

    static std::shared_ptr<JavaSerializable> createObject(const std::string &className)
    {
        auto it = creators.find(className);
        if (it == creators.end())
        {
            throw std::runtime_error("Unknown class: " + className);
        }
        return it->second();
    }
};

class JavaDeserializer
{
private:
    static std::map<int, std::shared_ptr<JavaSerializable>> deserializedObjects;
    static int objectCounter;

public:
    static std::shared_ptr<JavaSerializable> deserialize(const std::vector<uint8_t> &data)
    {
        deserializedObjects.clear();
        objectCounter = 0;
        ByteReader reader(data);
        return deserializeObject(reader);
    }

private:
    static std::shared_ptr<JavaSerializable> deserializeObject(ByteReader &reader)
    {
        uint8_t nullMarker = reader.readByte();
        if (nullMarker == 0)
        {
            return nullptr;
        }

        uint8_t referenceMarker = reader.readByte();
        if (referenceMarker == 1)
        {
            int objectId = reader.readInt();
            auto it = deserializedObjects.find(objectId);
            if (it == deserializedObjects.end())
            {
                throw std::runtime_error("Invalid object reference");
            }
            return it->second;
        }

        // Read class metadata
        std::string className = reader.readString();

        // Create object using factory
        auto obj = ObjectFactory::createObject(className);
        deserializedObjects[objectCounter++] = obj;

        // Read field count
        int fieldCount = reader.readInt();

        // Read field metadata
        std::vector<std::pair<std::string, std::string>> expectedFields = obj->getFieldMetadata();
        if (fieldCount != expectedFields.size())
        {
            throw std::runtime_error("Field count mismatch");
        }

        // Read and verify field metadata
        for (int i = 0; i < fieldCount; i++)
        {
            std::string fieldName = reader.readString();
            std::string fieldType = reader.readString();

            if (fieldName != expectedFields[i].first ||
                fieldType != expectedFields[i].second)
            {
                throw std::runtime_error("Field metadata mismatch");
            }
        }

        // Deserialize fields
        obj->deserializeFields(reader);
        return obj;
    }
};

// Initialize static members
std::map<int, std::shared_ptr<JavaSerializable>> JavaDeserializer::deserializedObjects;
int JavaDeserializer::objectCounter = 0;
std::map<std::string, std::function<std::shared_ptr<JavaSerializable>()>> ObjectFactory::creators;

// Initialize static members
std::map<const void *, int> JavaSerializer::serializedObjects;
int JavaSerializer::objectCounter = 0;

File: Java/utils/Serializer.java
package SC4051_dist_system_proj.utils;

import java.util.HashMap;
import java.util.Map;

public class Serializer {
    // Track objects already serialized to handle circular references
    private static Map<Object, Integer> serializedObjects = new HashMap<>();
    private static Map<Integer, Object> deserializedObjects = new HashMap<>();
    private static int objectCounter = 0;

    public static byte[] serialize(Object obj) throws Exception {
        serializedObjects.clear();
        objectCounter = 0;
        ByteBuffer buffer = new ByteBuffer(1024);
        serializeObject(obj, buffer);
        return buffer.getBuffer();
    }

    private static void serializeObject(Object obj, ByteBuffer buffer) throws Exception {
        /*
         * Structure of the byte array:
         * 
         * Null marker
         * Object reference marker (Checks if object has been serialised before)
         * Object reference/handle (consists of the entire object's fieldname, fieldtype
         * + data.)
         * 
         * In each object reference/handle:
         * ClassName, field_lengths, fieldname, fieldType,fieldVal
         */

        if (obj == null) {
            buffer.writeByte((byte) 0); // null marker
            return;
        }

        buffer.writeByte((byte) 1); // non-null marker

        // Check if object type was already serialized. Check if object handle exists.
        if (serializedObjects.containsKey(obj)) {
            buffer.writeByte((byte) 1); // reference marker, object handle exists.
            buffer.writeInt(serializedObjects.get(obj));
            return;
        }

        buffer.writeByte((byte) 0); // reference marker, object handle dont exist.
        serializedObjects.put(obj, objectCounter++);

        // Write class metadata
        buffer.writeString(obj.getClass().getName());

        // Issue with accessing restricted internal Java classes for wrapper classes
        // types Integer, Long, etc.,
        // which throws an error.
        java.lang.reflect.Field[] allFields = obj.getClass().getDeclaredFields();
        java.util.List<java.lang.reflect.Field> serializableFields = new java.util.ArrayList<>();

        for (java.lang.reflect.Field field : allFields) {
            // System.out.println("object's field: " + field);
            int modifiers = field.getModifiers();
            if (!java.lang.reflect.Modifier.isStatic(modifiers) &&
                    !java.lang.reflect.Modifier.isTransient(modifiers)) {
                serializableFields.add(field);
            }
        }

        // Write the number of fields to serialise.
        buffer.writeInt(serializableFields.size());

        for (java.lang.reflect.Field field : serializableFields) {
            field.setAccessible(true);

            // Write field metadata
            buffer.writeString(field.getName());
            buffer.writeString(field.getType().getName());

            // Write field value
            writeValue(buffer, field.get(obj), field.getType());
        }
    }

    private static void writeValue(ByteBuffer buffer, Object value, Class<?> type) throws Exception {
        if (type.isPrimitive()) {
            writePrimitive(buffer, value, type);
        } else if (type == String.class) {
            if (value == null) {
                buffer.writeByte((byte) 0);
            } else {
                buffer.writeByte((byte) 1);
                buffer.writeString((String) value);
            }
        } else if (type.isArray()) {
            writeArray(buffer, value, type);
        } else {
            // Handle nested object
            serializeObject(value, buffer);
        }
    }

    // TODO bug here! Integer has a bug, I suspect the rest too. Cant even compile.
    // private static void writePrimitive(ByteBuffer buffer, Object value, Class<?>
    // type) {
    // // System.out.println("detected primitive: " + type);
    // if (type == int.class) {
    // buffer.writeInt((Integer) value);
    // } else if (type == long.class) {
    // buffer.writeLong((Long) value);
    // } else if (type == double.class) {
    // buffer.writeDouble((Double) value);
    // } else if (type == boolean.class) {
    // buffer.writeByte((byte) ((Boolean) value ? 1 : 0));
    // }
    // }

    private static void writePrimitive(ByteBuffer buffer, Object value, Class<?> type) {
        if (type == int.class) {
            System.out.println("detected int: " + value);
            buffer.writeInt(((Number) value).intValue());
        } else if (type == long.class) {
            buffer.writeLong(((Number) value).longValue());
        } else if (type == double.class) {
            buffer.writeDouble(((Number) value).doubleValue());
        } else if (type == byte.class) {
            buffer.writeByte(((Number) value).byteValue());
        }
    }

    private static void writeArray(ByteBuffer buffer, Object value, Class<?> type) throws Exception {
        if (value == null) {
            buffer.writeInt(-1);
            return;
        }

        int length = java.lang.reflect.Array.getLength(value);
        buffer.writeInt(length);

        Class<?> componentType = type.getComponentType();
        for (int i = 0; i < length; i++) {
            Object element = java.lang.reflect.Array.get(value, i);
            writeValue(buffer, element, componentType);
        }
    }

    /*
     * Structure of the byte array:
     * 
     * Null marker
     * Object reference marker (Checks if object has been serialised before)
     * Object reference/handle (consists of the entire object's fieldname, fieldtype
     * + data.)
     * 
     * In each object reference/handle:
     * ClassName, field_lengths, fieldname, fieldType,fieldVal
     */

    public static Object deserialize(byte[] data) throws Exception {
        deserializedObjects.clear();
        objectCounter = 0;
        ByteReader reader = new ByteReader(data);
        return deserializeObject(reader);
    }

    /*
     * getClassFromName(String className) is a Helper function for
     * deserializeObject().
     * 
     * Primitive class types cannot get the class name using Class.forName method
     * during deserialisation
     */
    private static Class<?> getClassFromName(String className) throws ClassNotFoundException {
        // Handle primitive types
        switch (className) {
            case "int":
                return int.class;
            case "long":
                return long.class;
            case "double":
                return double.class;
            case "float":
                return float.class;
            case "boolean":
                return boolean.class;
            case "byte":
                return byte.class;
            case "short":
                return short.class;
            case "char":
                return char.class;
            case "void":
                return void.class;
            default:
                return Class.forName(className);
        }
    }

    private static Object deserializeObject(ByteReader reader) throws Exception {
        byte nullMarker = reader.readByte();
        if (nullMarker == 0) {
            return null;
        }

        // check if object has been deserialised before.
        byte referenceMarker = reader.readByte();
        if (referenceMarker == 1) {
            int objectId = reader.readInt();
            return deserializedObjects.get(objectId);
        }

        // Read class metadata
        String className = reader.readString();
        Class<?> cls = getClassFromName(className);

        System.out.println("object class: " + cls);

        // Handle primitive types directly
        if (cls.isPrimitive()) {
            System.out.println("detected primitive: " + cls);
            Object value = readPrimitive(reader, cls);
            deserializedObjects.put(objectCounter++, value);
            return value;
        }

        Object obj = cls.getDeclaredConstructor().newInstance();

        // Store object before deserializing fields to handle circular references
        deserializedObjects.put(objectCounter++, obj);

        int fieldCount = reader.readInt();
        for (int i = 0; i < fieldCount; i++) {
            String fieldName = reader.readString();
            String fieldTypeName = reader.readString();

            java.lang.reflect.Field field = cls.getDeclaredField(fieldName);

            // Skip if field is static or transient
            if (!java.lang.reflect.Modifier.isStatic(field.getModifiers()) &&
                    !java.lang.reflect.Modifier.isTransient(field.getModifiers())) {
                field.setAccessible(true);
                System.out.println("fieldtypename: " + fieldTypeName);
                Class<?> fieldType = getClassFromName(fieldTypeName);
                field.set(obj, readValue(reader, fieldType));
            }
        }

        return obj;
    }

    private static Object readValue(ByteReader reader, Class<?> type) throws Exception {
        if (type.isPrimitive()) {
            return readPrimitive(reader, type);
        } else if (type == String.class) {
            byte nullMarker = reader.readByte();
            return nullMarker == 0 ? null : reader.readString();
        } else if (type.isArray()) {
            return readArray(reader, type);
        } else {
            return deserializeObject(reader);
        }
    }

    private static Object readPrimitive(ByteReader reader, Class<?> type) {
        System.out.println("detected primitive: " + type);
        if (type == int.class) {
            return reader.readInt();
        } else if (type == long.class) {
            return reader.readLong();
        } else if (type == double.class) {
            return reader.readDouble();
        } else if (type == boolean.class) {
            return reader.readByte() == 1;
        }
        return null;
    }

    private static Object readArray(ByteReader reader, Class<?> type) throws Exception {
        int length = reader.readInt();
        if (length == -1) {
            return null;
        }

        Object array = java.lang.reflect.Array.newInstance(type.getComponentType(), length);
        for (int i = 0; i < length; i++) {
            java.lang.reflect.Array.set(array, i, readValue(reader, type.getComponentType()));
        }
        return array;
    }
}
